crates/bevy_render/src/renderer/mod.rs

```diff
@@ -345,25 +360,87 @@ impl RenderContext {
         TrackedRenderPass::new(&self.render_device, render_pass)
     }
 
-    /// Append a [`CommandBuffer`] to the queue.
+    /// Append a [`CommandBuffer`] to the command buffer queue.
     ///
     /// If present, this will flush the currently unflushed [`CommandEncoder`]
-    /// into a [`CommandBuffer`] into the queue before append the provided
+    /// into a [`CommandBuffer`] into the queue before appending the provided
     /// buffer.
     pub fn add_command_buffer(&mut self, command_buffer: CommandBuffer) {
         self.flush_encoder();
-        self.command_buffers.push(command_buffer);
+
+        self.command_buffer_queue
+            .push(QueuedCommandBuffer::Ready(command_buffer));
     }
 
-    /// Finalizes the queue and returns the queue of [`CommandBuffer`]s.
+    /// Append a function that will generate a [`CommandBuffer`] to the
+    /// command buffer queue, to be ran later.
+    ///
+    /// If present, this will flush the currently unflushed [`CommandEncoder`]
+    /// into a [`CommandBuffer`] into the queue before appending the provided
+    /// buffer.
+    pub fn add_command_buffer_generation_task(
+        &mut self,
+        task: impl FnOnce(RenderDevice) -> CommandBuffer + 'w + Send,
+    ) {
+        self.flush_encoder();
+
+        self.command_buffer_queue
+            .push(QueuedCommandBuffer::Task(Box::new(task)));
+    }
+
+    /// Finalizes and returns the queue of [`CommandBuffer`]s.
+    ///
+    /// This function will wait until all command buffer generation tasks are complete
+    /// by running them in parallel (where supported).
     pub fn finish(mut self) -> Vec<CommandBuffer> {
         self.flush_encoder();
-        self.command_buffers
+
+        if self.command_buffer_queue.is_empty() {
+            self.command_buffer_queue
+                .into_iter()
+                .map(|queued_command_buffer| match queued_command_buffer {
+                    QueuedCommandBuffer::Ready(command_buffer) => command_buffer,
+                    QueuedCommandBuffer::Task(_) => unreachable!(),
+                })
+                .collect()
+        } else {
+            let mut command_buffers = Vec::with_capacity(self.command_buffer_queue.len());
+            let mut task_based_command_buffers = ComputeTaskPool::get().scope(|task_pool| {
+                for (i, queued_command_buffer) in self.command_buffer_queue.into_iter().enumerate()
+                {
+                    match queued_command_buffer {
+                        QueuedCommandBuffer::Ready(command_buffer) => {
+                            command_buffers.push((i, command_buffer));
+                        }
+                        QueuedCommandBuffer::Task(command_buffer_generation_task) => {
```

Comment:
Is there a reason why we can't immediately launch the task on queuing it, and just collect the results in `finish`? As this stands, if we have a large blocking part of the graph, the actual parallel tasks won't get started until that part of the graph is complete.

Reply:
The tasks need to be scoped, which isn't easy to do unless we start them all at the end. For now, this should be fine. I don't foresee it really being a problem.